service-base/src/main/java/org/eclipse/hono/service/amqp/RequestResponseEndpoint.java

```diff
@@ -237,54 +296,179 @@ protected final void handleMessage(final ProtonConnection con, final ProtonRecei
 
-            sender.setQoS(ProtonQoS.AT_LEAST_ONCE);
-            HonoProtonHelper.setCloseHandler(sender, senderClosed -> {
-                logger.debug("client [{}] closed sender link, removing associated event bus consumer [{}]", sender.getName(), replyConsumer.address());
-                replyConsumer.unregister();
-                if (senderClosed.succeeded()) {
-                    senderClosed.result().close();
-                }
-            });
-            sender.open();
-        } else {
+        if (!isValidReplyToAddress(replyToAddress)) {
             logger.debug("client [{}] provided invalid reply-to address", sender.getName());
             sender.setCondition(ProtonHelper.condition(AmqpError.INVALID_FIELD,
-                    String.format("reply-to address must have the following format %s/<tenant>/<reply-address>", getName())));
+                    String.format("reply-to address must have the following format %s/<tenant>/<reply-address>",
+                            getName())));
             sender.close();
+            return;
+        }
+
+        final String replyTo = replyToAddress.toString();
+
+        if (this.replyAddresses.contains(replyTo)) {
+            logger.debug("client [{}] wanted to subscribe to already subscribed reply-to address [{}]",
+                    sender.getName(), replyTo);
+            sender.setCondition(ProtonHelper.condition(AmqpError.ILLEGAL_STATE,
+                    String.format("reply-to address [%s] must not be subscribed yet", replyTo)));
+            sender.close();
+            return;
+        }
+
+        logger.debug("establishing response sender link with client [{}]", sender.getName());
+
+        final MessageConsumer<JsonObject> replyConsumer = vertx.eventBus().consumer(replyTo,
+                message -> {
+                    // TODO check for correct session here...?
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("forwarding reply to client [{}]: {}", sender.getName(),
+                                message.body().encodePrettily());
+                    }
+                    final EventBusMessage response = EventBusMessage.fromJson(message.body());
+                    filterResponse(Constants.getClientPrincipal(con), response)
+                            .recover(t -> {
+                                final int status = ServiceInvocationException.extractStatusCode(t);
+                                return Future.succeededFuture(response.getResponse(status));
+                            })
+                            .map(filteredResponse -> {
+                                final Message amqpReply = getAmqpReply(filteredResponse);
+                                sender.send(amqpReply);
+                                flowCreditToRequestor(replyTo);
+                                return null;
+                            })
+                            .recover(t -> {
```

Comment:
in which case is this supposed to be executed?

Reply:
sure, but which method do you expect/suspect to fail there?

Reply:
If I would expect a method to fail then I would do it differently :grin: 
But yes, the main reason is `getAmqpReply`.

Reply:
My understanding is that you want to make sure that we flow credit to the sender even if we are not able to create a response message properly (i.e. getAmypReply fails), right? If that is the case then I would really prefer to put the call to *getAmqpReply* and sender.send() into a try/catch block. That would make the intent much clearer FMPOV

Reply:
Like that?